design/one-pager-scope-workload-interaction-mechanism.md

```diff
@@ -0,0 +1,141 @@
+# Scopes and workloads interaction mechanism in OAM
+
+* Owner: Ryan Zhang (@ryanzhang-oss)
+* Status: Implemented
+
+This proposal stays in line with [Traits and workloads interaction mechanism in OAM](https://github.com/crossplane/oam-kubernetes-runtime/blob/master/design/one-pager-trait-workload-interaction-mechanism.md) and enables scopes a similar mechanism with how traits interact with workloads.
+
+## Background
+> Application scopes are used to group components together into logical applications by providing different forms of application boundaries with common group behaviors.
+
+It's a crucial step for OAM implementation to make Scope aware of which workloads are associated to it. Currently, OAM implementation assumes that every Scope has a field `.spec.workloadRefs` to record associated workload references and get/set references in a hard-code way. It's in need of a more generic mechanism to do that.
+ 
+We will use the following hypothetical OAM application as the baseline to illustrate the problem and our solution. 
+
+`HealthScope` CRD refers to [here](https://github.com/crossplane/oam-kubernetes-runtime/blob/master/charts/oam-kubernetes-runtime/crds/core.oam.dev_healthscopes.yaml).
+ 
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: WorkloadDefinition
+metadata:
+  name: example-workload.oam.dev
+spec:
+  definitionRef:
+    name: example-workload.oam.dev
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: ScopeDefinition
+metadata:
+  name: healthscope.core.oam.dev
+spec:
+  allowComponentOverlap: true
+  definitionRef:
+    name: healthscope.core.oam.dev
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: HealthScope
+metadata:
+  name: example-health-scope
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: example-component
+spec:
+  workload:
+    apiVersion: oam.dev/v1alpha2
+    kind: example-workload
+    metadata: 
+      name: my-example-workload
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: example-component      
+      scopes:
+        - scopeRef:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: HealthScope
+            name: example-health-scope
+```
+
+The problem is two folds
+1. A scope controller needs a way to find the workloads that it's going to control. 
+   - In the example, HealthScope should know that it is supposed to check the health condition of `my-example-workload`. 
+However, we want to keep the applicationConfiguration controller agnostic to the schema of any `scope` it interacts with or `workload` it generates to make it extensible. 
+Thus, the applicationConfiguration controller needs to get the HealthScope CR (because scope is a kind of global resource) and give it a reference of the `my-example-workload` workload **without knowing HealthScope's specific schema**.
```

Comment:
We have already implemented `workloadRefs` in scope which solves this problem, I think we should narrow the problem to be `current implementation is hard code` issue.